gobblin-runtime/src/main/java/gobblin/runtime/Fork.java

```diff
@@ -171,30 +309,57 @@ public void close()
   }
 
   /**
+   * Get new records off the record queue and process them.
+   */
+  private void processRecords() throws IOException, DataConversionException {
+    while (true) {
+      Object record;
+      try {
+        record = this.recordQueue.get();
+      } catch (InterruptedException ie) {
+        this.logger.warn("Interrupted while trying to get a record off the queue", ie);
+        continue;
+      }
+      if (record == null) {
+        // The parent task has already done pulling records so no new record means this fork is done
+        if (this.parentTaskDone) {
+          return;
+        }
+      } else {
+        // Convert the record, check its data quality, and finally write it out if quality checking passes.
+        for (Object convertedRecord : this.converter.convertRecord(this.convertedSchema, record, this.taskState)) {
+          if (this.rowLevelPolicyChecker.executePolicies(convertedRecord, this.rowLevelPolicyCheckingResult)) {
+            this.writer.write(convertedRecord);
+          }
+        }
+      }
+    }
+  }
+
+  /**
    * Check data quality.
    *
    * @return whether data publishing is successful and data should be committed
    */
-  private boolean checkDataQuality(long recordsPulled, long expectedCount, long pullLimit, Object schema)
+  private boolean checkDataQuality(Object schema)
       throws Exception {
-
-    if (pullLimit > 0) {
-      // If pull limit is set, use the actual number of records pulled.
-      this.taskState.setProp(ConfigurationKeys.EXTRACTOR_ROWS_EXPECTED, recordsPulled);
-    } else {
-      // Otherwise use the expected record count
-      this.taskState.setProp(ConfigurationKeys.EXTRACTOR_ROWS_EXPECTED, expectedCount);
+    TaskState taskStateForFork = this.taskState;
+    if (this.branches > 1) {
+      // Make a copy if there are more than one fork
+      taskStateForFork = new TaskState(this.taskState);
```

Comment:
I know this is an existing API but it looks very confusing/misleading to me. "new TaskState(this.taskState)" looks like a copy constructor but it is not since it does not copy the properties. Then we have a separate addAll() to copy the properties which also changes the workunit property along the way. Is there a use case to copy a task without its properties, i.e. should addAll() be merged in TaskState(this.taskState)?